### PR TITLE
[dpdk] Eliminate header copy

### PR DESCRIPTION
**Pull request by kamleshbhalui**
Dpdk does not support direct header copy.
So, Eliminate temporary copies of header, which generally populated after inlining, if it;s not temporary copy then transform direct header copy to element wise copy.

### CODE REVIEW

**Review by usha1830 — APPROVED** [review]
Looks fine.
Please move the function definitions to .cpp

File: backends/dpdk/dpdkArch.h [hunk 1]
```diff
@@ -926,6 +926,70 @@ class CollectErrors : public Inspector {
...
+                                            new IR::Member(as->left, field->name),
+                                            new IR::Member(as->right, field->name)));
+                }
+                return new IR::BlockStatement(as->srcInfo, components);
+            }
+        }
+        return as;
+    }
+
+    const IR::Node* preorder(IR::MethodCallExpression *me) override {
```
[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.h:964]: it's for taking care of compiler generated cases like , eth_0.setInvalid();

[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.h:964]: here also eth_0 needs to be replaced with hdr.ethernet

File: backends/dpdk/dpdkArch.h [hunk 2]
```diff
@@ -926,6 +926,70 @@ class CollectErrors : public Inspector {
...
 
+class ElimHeaderCopy : public Transform {
+    P4::TypeMap *typeMap;
+    ordered_map<cstring, const IR::Member*> replacementMap;
+
+ public:
+    explicit ElimHeaderCopy(P4::TypeMap *typeMap) : typeMap{typeMap} {}
+    bool isHeader(const IR::Expression* e) {
+        auto type = typeMap->getType(e);
+        return type->is<IR::Type_Header>() && !e->is<IR::MethodCallExpression>();
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkArch.h:937]: it's for ignoring cases like option_t tmp_hdr = pkt.lookahead<option_t>();

File: backends/dpdk/dpdkArch.cpp [hunk 3]
```diff
@@ -2253,4 +2253,61 @@ void CollectAddOnMissTable::postorder(const IR::MethodCallStatement *mcs) {
...
+        return me;
+    if (!isHeader(m->expr))
+        return me;
+    auto expr = m->expr->to<IR::PathExpression>();
+    if (expr && replacementMap.count(expr->path->name.name)) {
+        auto me1 = new IR::MethodCallExpression(me->srcInfo,
+                     new IR::Member(replacementMap.at(expr->path->name.name), m->member),
+                     me->typeArguments, me->arguments);
+    typeMap->setType(me1->method, typeMap->getType(me->method));
+    return me1;
```
Comment (github-sajan) [inline backends/dpdk/dpdkArch.cpp:2297]: indentation

File: backends/dpdk/dpdkArch.h [hunk 4]
```diff
@@ -926,6 +926,18 @@ class CollectErrors : public Inspector {
     }
 };
```
Comment (github-sajan) [inline backends/dpdk/dpdkArch.h:928]: Please add some description for this class usage

**Review by github-sajan — APPROVED** [review]
Looks fine

**Pull request closed without merging**